lib/Image/HtmlToImage.php

```diff
@@ -0,0 +1,205 @@
+            return self::$supportedAdapter;
+        }
+
+        self::$supportedAdapter = '';
+
+        if (GotenbergHelper::isAvailable()) {
+            /** @var GotenbergAPI|object $chrome */
+            $chrome = GotenbergAPI::chromium(Config::getSystemConfiguration('gotenberg')['base_url']);
+            if(method_exists($chrome, 'screenshot')) {
+                // only v2 of Gotenberg lib is supported
+                self::$supportedAdapter = 'gotenberg';
+            }
+        }
+
+        if (!self::$supportedAdapter && class_exists(BrowserFactory::class)) {
+            $chromiumUri = \Pimcore\Config::getSystemConfiguration('chromium')['uri'];
+            if (!empty($chromiumUri)) {
+                try {
+                    if((new Connection($chromiumUri))->connect()) {
+                        self::$supportedAdapter = 'chromium';
+                    }
+                } catch (\Exception $e) {
+                    Logger::debug((string) $e);
+                    // nothing to do
+                }
+            }
+
+            if(self::getChromiumBinary()) {
+                self::$supportedAdapter = 'chromium';
+            }
+        }
+
+        return self::$supportedAdapter;
+    }
+
+    public static function getChromiumBinary(): ?string
+    {
+        foreach (['chromium', 'chrome'] as $app) {
+            $chromium = Console::getExecutable($app);
+            if ($chromium) {
+                return $chromium;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public static function convert(string $url, string $outputFile, ?string $sessionName = null, ?string $sessionId = null, string $windowSize = '1280,1024'): bool
+    {
+        $adapter = self::getSupportedAdapter();
+        if($adapter === 'gotenberg') {
+            return self::convertGotenberg(...func_get_args());
+        } elseif ($adapter === 'chromium') {
+            return self::convertChromium(...func_get_args());
+        }
+
+        return false;
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public static function convertGotenberg(string $url, string $outputFile, ?string $sessionName = null, ?string $sessionId = null, string $windowSize = '1280,1024'): bool
+    {
+        try {
+
+            $extraHeaders = [
+                'X-Foo' => 'Bar' // required, as extraHttpHeaders() requires at least one entry
+            ];
+
+            if (null !== $sessionId && null !== $sessionName) {
+                $extraHeaders['Cookie'] = $sessionName . '=' . $sessionId;
+            }
+
+            /** @var GotenbergAPI|object $request */
+            $request = GotenbergAPI::chromium(Config::getSystemConfiguration('gotenberg')['base_url']);
+            if(method_exists($request, 'screenshot')) {
+                $request->screenshot()
+                    ->png()
+                    ->extraHttpHeaders($extraHeaders)
+                    ->url($url);
```

Comment:
```suggestion
                    ->html($html);
```
Not the actual fix, for the comparison between 2 historical version.

The problem is that `$url` is passing `localhost` while the gotenberg container to work will need to be `nginx:80` service to connect in-between.

It's not simple as replacing the domain though, in web-to-print bundle we rely on [Mail::setAbsolutePaths](https://github.com/pimcore/web-to-print-bundle/blob/fb647d037a51f3d50f1e86ef4e2b224c8befe4e0/src/Processor.php#L281)
to change all the link/css/js to be [nginx:80](https://github.com/pimcore/web-to-print-bundle/blob/fb647d037a51f3d50f1e86ef4e2b224c8befe4e0/src/Processor/Gotenberg.php#L43) to work.

Another challenge here is that we need admin rights/access to use the `?pimcore_version=110`

Reply:
@kingjia90 There's the setting: 
`pimcore.documents.preview_url_prefix`, see also in docs: 
https://pimcore.com/docs/platform/Pimcore/Installation_and_Upgrade/System_Setup_and_Hosting/Additional_Tools_Installation/#gotenberg 

There you can set `nginx:80` 😊 

Authentication is done via sessions, that's why the session get's passed here: 
https://github.com/pimcore/admin-ui-classic-bundle/blob/9258d42920dbb475badc1adea59a7552ab089ac4/src/Controller/Admin/Document/DocumentController.php#L1041

Reply:
ah, thank you, i guess i was trying statically and missed that ..
but does it work for you? i get blank images (white background)

Reply:
@kingjia90 yes, it used to work for me, strange 🤔

Reply:
somehow it throws `URL screenshot: screenshot: context deadline exceeded` in the Gotenberg container log, but any other website eg. `pimcore.com` would work if i change the url 😟 no any clues on what is happening as it quite tricky to debug

Reply:
@kingjia90 I also have `URL screenshot: screenshot: context deadline exceeded`in logs but still I'm getting a screenshot rendered. 

But the strange thing is that I'm not getting 2 screenshots in a row, for version comparison in document pages. 
And even stranger: it seems that generating the version diff even doesn't work with current implementation, see demo.pimcore.fun (tested with `/en/Magazine/Culture`).

Reply:
Ok, now after restarting the Gotenberg container also multiple requests in a row work as well - very strange...

Reply:
Maybe one thing to consider is that for demo you need to customize the startup of Gotenberg: 

in `docker-compose.override.yaml`: 
```yaml
services: 
    gotenberg:
        image: gotenberg/gotenberg:8
        command: [ gotenberg, --chromium-auto-start=true, --chromium-ignore-certificate-errors=true, --chromium-disable-web-security=true]
```

Reply:
Is the screenshot rendered with the historical changes highlighted in yellow? Asking because i suspect it is rendering only the first one but "timeouts" on the second document

The input that current implementation is not properly working is a nice starting point, i didn't consider that and will check that as well. 
Maybe there is some infinite redirect, because i could see the screenshot with errors if i  fatal break something in "public/index.php"

Reply:
So imho the Gotenberg implementation itself seems to be fine, but we have an issue with this here: 
https://github.com/pimcore/admin-ui-classic-bundle/blob/9258d42920dbb475badc1adea59a7552ab089ac4/src/Controller/Admin/Document/DocumentController.php#L1041-L1042 

As it seems that just passing the `Cookie` header isn't enough anymore. 
I'd assume we need to render the HTML using `Pimcore\Document\Renderer\DocumentRenderer` then inject a `<base href="{{ pimcore.documents.preview_url_prefix }}" />` into the rendered HTML and then pass it to `->html()` method of Gotenberg.

Reply:
@kingjia90 no, changes are not highlighted, as it seems that using the `Cookie` auth workaround doesn't work anymore. 
For me also the comparison doesn't work on current demo.pimcore.fun -> so I'd assume it a problem in general.

Reply:
Demo probably broken due https://github.com/pimcore/pimcore/pull/16765 and/or https://github.com/pimcore/demo/pull/571, also it seems locally that the `preview_url_prefix` should be prefixed with the protocol as `http://nginx:80` instead of `nginx:80`